### PR TITLE
*Edit update relation data when data is removed in client side

### PR DESCRIPTION
**Pull request by wongpratan**

### CODE REVIEW

**Comment by Hiro-Nakamura** [comment]
This means that the DB entry is deleted automatically, but the Client hasn't been notified about that.  This causes the Client to still display and let you work with an entry that isn't there.

So first thing we need to do is when an Object Entry is deleted, we need to scan each of it's connections, and find a connection to another Object that is a 1:x relations.  For each of those, we need to issue a **'ab.datacollection.delete'** socket message as well.

**Comment by Hiro-Nakamura** [comment]
2nd:  we also want to make this behavior more controllable.  For example, in the case of Events : Registration ,  if we delete an Event then we DO want to remove the registrations.  However in the case of a  Staff : StaffType,  if we delete a StaffType entry, we DO NOT want to remove the Staff assigned to that.  

So we talked about adding another checkbox on the Add Connected Object  dialog, to indicate if we should `On Delete Cascade` or `On Delete Set Null`.

screen shot 2018-12-12 at 3 39 50 pm

If the checkbox is checked then `On Delete Cascade`.  Otherwise `On Delete Set Null`

We would only show the checkbox option if one of the entries are `One` type.  If they are `Many:Many`, then we would not show the checkbox (and default to `On Delete Set Null` ).

**Comment by Hiro-Nakamura** [comment]
3rd, after we do those, let's explore what it will take to modify the `UPDATE` process to also reduce the number of `.stale` notifications as well.

**Comment by wongpratan** [comment]
> This means that the DB entry is deleted automatically, but the Client hasn't been notified about that. This causes the Client to still display and let you work with an entry that isn't there.
> 
> So first thing we need to do is when an Object Entry is deleted, we need to scan each of it's connections, and find a connection to another Object that is a 1:x relations. For each of those, we need to issue a **'ab.datacollection.delete'** socket message as well.

Yes, I think we have a more option.
If we unrelate data before delete, then we can keep data and reduce 'ab.datacollection.delete' messages.

**Comment by wongpratan** [comment]
> OK, on 1st review, the immediate Delete of an entry is getting properly updated across the connected objects that linked to that object. That is good.
> 
> However, we are noticing an additional issue with deleting items. When an object is defined as a 1:many relationship with another object, it seems like the Object with the 1 constraint is being Deleted when the many object is deleted.
> 
> For example, in the Events App: Registration : Event is 1:many. And the Registration obj is defaulted to DELETE CASCADE:
> screen shot 2018-12-12 at 3 15 50 pm

It may come from the old version of code.
Now, It does not set ON DELETE CASCADE. It is ON DELETE SET NULL.

https://github.com/appdevdesigns/app_builder/commit/27d4fbd317f52f9ddc582bc74b2482ff0f0a8f53#diff-2681e43b45ad3d5bff2052d433750eb3

File: assets/opstools/AppBuilder/classes/views/ABViewDataCollection.js [hunk 1]
```diff
@@ -1073,35 +1073,111 @@ export default class ABViewDataCollection extends ABView {
...
+			if (obj.id == data.objectId) {
 
-			if (this.__dataCollection.exists(values.id)) {
-				// normalize data before update data collection
-				var model = obj.model();
-				model.normalizeData(values);
-				this.__dataCollection.updateItem(values.id, values);
+				// various PK name
+				if (!values.id && obj.PK() != 'id')
+					values.id = values[obj.PK()];
```
[the remark this answers is not in the document]
Comment (wongpratan) [reply assets/opstools/AppBuilder/classes/views/ABViewDataCollection.js]: webix's datatable and data collection need .id

**Pull request merged**